# SeqApiPop_WGShoneybeeDataReanalysis
Reanalysis of the WGS bee (SeqApiPop data) to estimate nucleotide diversity, load and infer demography

# 1/ Generate a complete VCF (./Pipeline2CompleteVCF)
We used the g.vcf files generated by Wragg and collaborators (2022), for details regarding these previous steps (mapping, g.vcf, etc), see https://github.com/avignal5/SeqApiPop <br>
To speed up the combination, all scripts were run independently for each chromosome (see script_lanceur_combinevcf_chrperchr.sh, this script launches script_combinevcf_Dorsata.sh for each of the chromosomes, thanks to the information provided by on the file HAv3_1_Chromosomes.list). Note that the example provided is for an outgroup (Dorsata), but the same strategy was used for the focal species (SeqApiPop data, see also script_combinevcf_870SeqApiPop.sh) and the two other outgroups. <br><br>
Then, the jointgenotyping was perfomed using GenotypeGVCFs (GATK v.4.2.2.0) using the script script_lanceur_genotypegvcf_chrperchr.sh (see also script_genotypeGVCFsHAV3_1.sh). Note that we use here the "--all-sites true" mode, in order to generate a complete VCF (i.e. a VCF with both variant and non-variant positions). <br>

# 2/ Reconstruct fasta sequences from a complete vcf (./fasta_sequence_reconstruction)
To generate perfectly aligned fasta sequences, we considered the approach developped by Leroy et al. 2021 (Current Biology), adjusting for the level of ploidy (haploid for the A. mellifera samples, diploid for the other species). This step makes assumptions. First, this approach ignores INDELs and SVs in order to keep the sequences perfectly aligned. So, the diversity that is reported in our analysis corresponds to the diversity at SNPs. Second, the reconstructed sequences should not be considered as phased. Of course, by reconstructing the sequences of haploid individuals, the sequences are by definition phased, but for the outgroups, the sequences should be considered as unphased. In our analysis, no summary statistics that were used are based on the pattern of linkage disequilibrium (e.g. DILS, nucleotide diversity, etc). Nevertheless, the reconstructed sequences SHOULD NOT BE USED as inputs for methods that require to know the haplotype phasing.
