# SeqApiPop_WGShoneybeeDataReanalysis
Reanalysis of the WGS bee (SeqApiPop data) to estimate nucleotide diversity, load and infer demography

# 1/ Generate a complete VCF (./Pipeline2CompleteVCF)
We used the g.vcf files generated by Wragg and collaborators (2022), for details regarding these previous steps (mapping, g.vcf, etc), see https://github.com/avignal5/SeqApiPop <br>
To speed up the combination, all scripts were run independently for each chromosome (see script_lanceur_combinevcf_chrperchr.sh, this script launches script_combinevcf_Dorsata.sh for each of the chromosomes, thanks to the information provided by on the file HAv3_1_Chromosomes.list). Note that the example provided is for an outgroup (Dorsata), but the same strategy was used for the focal species (SeqApiPop data, see also script_combinevcf_870SeqApiPop.sh). <br><br>
Then, the jointgenotyping was perfomed using GenotypeGVCFs (GATK v.4.2.2.0) using the script script_lanceur_genotypegvcf_chrperchr.sh (see also script_genotypeGVCFsHAV3_1.sh). Note that we use here the "--all-sites true" mode, in order to generate a complete VCF. <br>
