## Conserved heterogeneous landscapes of diversity in honeybees
This script provides general information about the rationale, as well as the scripts used for the reanalysis of the whole-genome sequence data from hundreds of individuals from the SeqApiPop data (<i>Apis mellifera</i>, Wragg et al. 2022), along with data from 3 other <i>Apis</i> species (<i>A. cerana</i>, Chen et al. 2018; <i>A. dorsata</i> & <i>A. laboriosa</i>, Cao et al. 2023). Overal, our objective was to estimate nucleotide diversity of honeybees, evaluate the conservation of their genomic landscapes and infer their demography history. <br>
Note that the associated manuscript is currently under preparation. <br>
Please send an email to <a href="mailto:thibault.leroy@inrae.fr?subject=[SeqApiPopReanalysis-Github]">me </a> for questions regarding these scripts.

### 1/ Generate a complete VCF (./Pipeline2CompleteVCF)
We used the g.vcf files generated by Wragg and collaborators (2022), for details regarding these previous steps (mapping, g.vcf, etc), see https://github.com/avignal5/SeqApiPop <br>
To speed up the combination, all scripts were run independently for each chromosome (see script_lanceur_combinevcf_chrperchr.sh, this script launches script_combinevcf_Dorsata.sh for each of the chromosomes, thanks to the information provided by on the file HAv3_1_Chromosomes.list). Note that the example provided is for an outgroup (Dorsata), but the same strategy was used for the focal species (SeqApiPop data, see also script_combinevcf_870SeqApiPop.sh) and the two other outgroups. <br><br>
Then, the jointgenotyping was perfomed using GenotypeGVCFs (GATK v.4.2.2.0) using the script script_lanceur_genotypegvcf_chrperchr.sh (see also script_genotypeGVCFsHAV3_1.sh). Note that we use here the "--all-sites true" mode, in order to generate a complete VCF (i.e. a VCF with both variant and non-variant positions). <br>

### 2/ Reconstruct fasta sequences from a complete vcf (./fasta_sequence_reconstruction)
Important note: To generate perfectly aligned fasta sequences, we considered the approach developped by Leroy et al. 2021 (Current Biology), adjusting for the level of ploidy (haploid for the A. mellifera samples, diploid for the other species). This step makes several assumptions. First, this approach ignores INDELs and SVs in order to keep the sequences perfectly aligned. So, the diversity that we report in our analysis corresponds to the diversity at SNPs. Second, the reconstructed sequences should not be considered as phased. Of course, given that the vast majority of our dataset corresponds to A. mellifera drones, these sequences are by definition phased. But it should be kept in mind that the sequences of the outgroup species are based on several workes and therefore, the reconstructed sequences should be considered as unphased. In our analysis, the analysis we performed are based on summary statistics that do not take linkage disequilibrium into account (e.g. DILS, nucleotide diversity, etc). So to be completely clear, the reconstructed sequences from diploid individuals SHOULD NOT BE USED as inputs for methods that require to know the haplotype phasing.<br><br>
At each position, the script will extend the sequence, by either adding a reference allele (non-polymorphic sites or SNPs with reference alleles), an alternate allele (SNPs with alternate alleles) or a missing information (N). To generate unbiased estimates of nucleotide diversity, it is indeed crucial to use information from invariant sites from the complete VCF, allowing to discriminate positions that are trully invariant from those that are just not covered (missing genotypes, e.g. for more information regarding this point, see the Pixy paper by Korunes & Samuk 2021 in Mol Ecol Res).<br>

To consider whether the position is sufficiently covered to be confident with the nucleotide to be added in the sequence, we first picked up 0.3% of all positions of the honeybee genome (~66k positions) to compute the quantiles of the coverage for each individual (see script_lanceur_quantilescoverage.sh). All positions that were not within the core of the distribution (excluding the bottom 5% and the top 5%) were hardmasked. A minimum coverage of 3 were required (it should be kept in mind that our dataset is almost exclusively based on haploid individuals). Then, we reconstructed the sequences using the python script "VCF2Fasta_haploid_withquantiles.py" (or its diploid equivalent "VCF2Fasta_diploid_withquantiles.py").<br>

### 3/ Generate 100kb blocks of sequences spanning the whole genome and compute summary statistics(./Sequence_Blocks_Nucleotide_Diversity)
From the chromosome-level sequences that we reconstructed, we then used bedtools (v.2-27.1) to generate block of sequences spanning the whole genome using bedtools getfasta, which were then used to subsequently generate the summary statistics of nucleotide diversity, Tajima's D etc using the software "seq_stat", see Leroy et al. 2021 Peer Community Journal. All these steps are listed in the script "script_compute_div_slidingwindows.sh", a script which is launched thanks to the "script_lanceur_div_slidingwindows.sh". <br>

For CDS regions, the approach is based on a similar strategy (see script_sumstats_diversity_CDS.sh), but requires different softwares: the python script cutSeqGff.py instead of bedtools getfasta, a small program called removeLastStopCodon and a dedicated software for the computation of the summary statistics "seq_stat_coding"), see also Leroy et al. 2021 Current Biology. The small programs cleanAlignment, removeLastStopCodon, seq_stat and seq_stat_coding are already compilated.

### 4/ Perform demographic reconstruction with DILS (./DILS_analysis)
The demographic approach we used is the one implemented in DILS, an ABC-based demographic modelling that can take into account the impacts of heterogenous landscapes of Ne (background selection, sweeps) and migration rates (reproductive barriers). We have therefore preferred the use of DILS among other methods, since the heterogeneities of Ne and Nm are known to potentially impact the accuracy of demographic modeling approches. For more information, see the DILS paper (Fraisse et al. 2021 Mol Ecol Res) and the associated github repository (https://github.com/dils-popgen/dils) <br>
For this study, we have generated a series of bash scripts that generate the inputfiles for DILS by selecting sequence data for pairs of individuals from a given number of 100kb windows (200 windows for the whole-genome data see "script_generate_inputfiles_DILS3_with_outgroup.sh", 50 windows for the scripts focusing only on non-rembining and recombining regions, see "script_generate_inputfiles_DILS3_with_outgroup_non-recomb.sh" and "script_generate_inputfiles_DILS3_with_outgroup_recomb.sh", respectively). An example of the input files needed, as well as the output files generated par DILS, are shown in the repository "./DILS_analysis/EXAMPLE_Mellifera_Carnica_constantNe_noSFS_outgroupLaboriosa". Note that only the first 200 lines of the file Sequences_Mellifera_Carnica_outLaboriosa_randomlyselected_200_genomic_regions.txt are shown in the file "Sequences_Mellifera_Carnica_outLaboriosa_randomlyselected_200_genomic_regions.head200.txt" because the complete file is very large (2 Gb). It is however sufficient to understand the requirements for the input files. <br>

A simple bash script was also used to parse the output of different runs of DILS (tar.gz files), summarizing the options used for the simulation, the best models and their posterior probabilities, as well as the the values of the main parameters (see script_parsing_output_DILS.sh). <br>
