# TL - 140318
# e.g. 
vcffile=$(echo "$1") # filtered vcf file (expected to be generated by the pipeline)
scaffold=$(echo "$vcffile" | sed 's/seqapipop870_jointgvcf_260123_//g' | sed 's/\.filtered\.Mellifera\.vcf//g' )
scaffoldlength=$(grep "$scaffold" ../GCF_003254395.2_Amel_HAv3.1_genomic.fna.scafflength | awk '{print $2}')
directory=$(echo "$2")
windowsize=$(echo "$3" | bc)
outprefix=$(echo "$4")
rm -r $outprefix.bed.tmp
mkdir $outprefix.bed.tmp
rm  $outprefix.$windowsize.bed
rm list_running_seq.txt

### generate block of sequences
echo "$scaffoldlength"
nbwindows=$(echo "($scaffoldlength / $windowsize ) + 1" | bc)
start=$(echo "1" )
echo "$scaffold	$nbwindows"
end=$(echo "$windowsize")
for i in $(eval echo "{1..$nbwindows}"); do
    # generate a bed
    rm ./$outprefix.bed.tmp/$outprefix.$windowsize.tmp
    while read line; do # for each individual, print a line in a tmp bed file
         echo "$line $start  $end" | sed 's/ \+/\t/g' >> ./$outprefix.bed.tmp/$outprefix.$windowsize.tmp
    done < sample_file.Mellifera.finallist.txt
    # keep the information in a sumup bed file
    echo "$scaffold $start  $end" | sed 's/ \+/\t/g' >> $outprefix.$windowsize.bed
    # bedtools getfasta
    rm  $directory/$scaffold.fst.fai # rm to reinitiate the fai
    bedtools getfasta -fi $directory/$scaffold.fst -bed ./$outprefix.bed.tmp/$outprefix.$windowsize.tmp -fo ./$outprefix.bed.tmp/$outprefix.$scaffold.$start.$end.fasta
    echo "$outprefix.$scaffold.$start.$end.fasta" >> list_running_seq.txt    
    # shift window
    previousend=$(echo "$end")
    start=$(echo "$end + 1" | bc)
    end=$(echo "$previousend + $windowsize" | bc)
done

### clean the alignments
cd ./$outprefix.bed.tmp/
find . -size 0 -delete # rm zero fil
ls *.fasta > list_withextracted_sequences
/work/genphyse/cytogen/Thibault/SeqApiPop_completevcf/cleanAlignment -seq list_withextracted_sequences -f fasta -n 3  -ploidy diploid # probably means that it requires 6 haploid individuals

### compute the summary statistics
for i in *.fasta.clean.fst; do 
    /work/genphyse/cytogen/Thibault/SeqApiPop_completevcf/seq_stat -seq $i -f fasta -o tmpstat
    if [ -f $outprefix.SlidWin$windowsize.stats ]; then
        tail -1 tmpstat >> $outprefix.SlidWin$windowsize.stats # other windows, print only stats
    else
        head -2 tmpstat > $outprefix.SlidWin$windowsize.stats  # first window, print header + stats
    fi
done
cd ..





